### PR TITLE
[BACKLOG-39678] - Schedule Output Perspective : Improve reporting engine to generate in line style and images for html rendering

### PR DESCRIPTION
**Pull request by soagarwal1**
@pentaho/hoth Please review

Part of PR set
https://github.com/pentaho/pentaho-scheduler-plugin/pull/179
https://github.com/pentaho/pentaho-scheduler-plugin-ee/pull/244
https://github.com/pentaho/pentaho-platform-plugin-reporting/pull/852

### CODE REVIEW

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 9m 17s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl engine/core
```
---
:ok_hand: All tests passed!
 
Tests run: **822**, Failures: **0**, Skipped: **2** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/reporting/job/PR-1657/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 9m 51s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl engine/core
```
---
:ok_hand: All tests passed!
 
Tests run: **822**, Failures: **0**, Skipped: **2** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/reporting/job/PR-1657/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 9m 49s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl engine/core
```
---
:ok_hand: All tests passed!
 
Tests run: **822**, Failures: **0**, Skipped: **2** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/reporting/job/PR-1657/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by singletonc — APPROVED** [review]

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 1]
```diff
@@ -544,6 +555,28 @@ private boolean tryHandleDrawable( final ReportAttributeMap attrs, final long wi
     return true;
   }
 
+  private void convertImageDataToBase64( String name, AttributeList attrList, final ImageContainer image, final String type,
+    final float quality ) throws ContentIOException {
+    int indexStart = name.lastIndexOf( "picture" );
+    String imgAltText = name.substring( indexStart );
+    attrList.setAttribute( HtmlPrinter.XHTML_NAMESPACE, ALT_ATTR, imgAltText );
+    attrList.setAttribute( HtmlPrinter.XHTML_NAMESPACE, SRC_ATTR, "poc-src-placeholder-" + imgAltText );
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:563]: you need to remove this line it clearly has "poc" "proof of concept" in it.

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 2]
```diff
@@ -544,6 +555,28 @@ private boolean tryHandleDrawable( final ReportAttributeMap attrs, final long wi
...
+    try {
+      final DefaultHtmlContentGenerator.ImageData imageData =
+        contentGenerator.getImageData( image, type, quality, true );
+      final String HTML_IMG_CSS_BASE64_FORMAT = "url('data:%1$s;base64,%2$s')";
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:567]: ` HTML_IMG_CSS_BASE64_FORMAT` needs to be a static final variable.

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 3]
```diff
@@ -544,6 +555,28 @@ private boolean tryHandleDrawable( final ReportAttributeMap attrs, final long wi
...
+      String urlBase64Encoding = String.format( HTML_IMG_CSS_BASE64_FORMAT,
+        imageData.getMimeType(), getBase64( imageData ) );
+      attrList.setAttribute( HtmlPrinter.XHTML_NAMESPACE, "content", urlBase64Encoding );
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:570]: string "content" needs to be replaced by newly added StyleBuilder.java enums  `CSSKeys#CONTENT` directly or `CSSKeys#CONTENT#getCssName()`

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 4]
```diff
@@ -661,6 +695,11 @@ private void writeImageTag( final StyleSheet styleSheet, final long width, final
       xmlWriter.writeTag( HtmlPrinter.XHTML_NAMESPACE, IMG_TAG, attrList, XmlWriterSupport.CLOSE );
       xmlWriter.writeCloseTag();
     } else {
+      String urlBase64Encoding  = attrList.getAttribute( HtmlPrinter.XHTML_NAMESPACE, "content", null );
+      if ( urlBase64Encoding != null ) {
+        imgStyle.append( StyleBuilder.CSSKeys.CONTENT, urlBase64Encoding );
+        attrList.removeAttribute( HtmlPrinter.XHTML_NAMESPACE, "content" );
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:701]: same comment here: https://github.com/pentaho/pentaho-reporting/pull/1657/files#r1626730575

don't repeat strings.

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 5]
```diff
@@ -544,6 +555,28 @@ private boolean tryHandleDrawable( final ReportAttributeMap attrs, final long wi
     return true;
   }
 
+  private void convertImageDataToBase64( String name, AttributeList attrList, final ImageContainer image, final String type,
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:558]: javadoc for HtmlTextExtractorHelper#convertImageDataToBase64(

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 6]
```diff
@@ -544,6 +555,28 @@ private boolean tryHandleDrawable( final ReportAttributeMap attrs, final long wi
...
+      final String HTML_IMG_CSS_BASE64_FORMAT = "url('data:%1$s;base64,%2$s')";
+      String urlBase64Encoding = String.format( HTML_IMG_CSS_BASE64_FORMAT,
+        imageData.getMimeType(), getBase64( imageData ) );
+      attrList.setAttribute( HtmlPrinter.XHTML_NAMESPACE, "content", urlBase64Encoding );
+    } catch ( Exception e ) {
+      throw new ContentIOException( "Conversion to Base64 failed", e );
+    }
+  }
+
+  public String getBase64( DefaultHtmlContentGenerator.ImageData imageData ) {
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:576]: javadoc for `HtmlTextExtractorHelper#getBase64`

Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:576]: lower the access level on javadoc for `HtmlTextExtractorHelper#getBase64`, I don't think it needs to be public. protected or lower will do.

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java [hunk 7]
```diff
@@ -496,7 +503,8 @@ private boolean tryHandleRotatedText( final RotatedTextDrawable rotatedTextDrawa
   }
 
   private boolean tryHandleDrawable( final ReportAttributeMap attrs, final long width, final long height,
```
Comment (NJtwentyone) [inline engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java:505]: javadoc HtmlTextExtractorHelper.java#tryHandleDrawable. More so explain what the new argument `boolean convertToBase64`  does. A developer shouldn't have to dig through the code to get a high level explanation of what is going on. **You are not responsible for the existing arguments.**

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
minor changes, only stopper is https://github.com/pentaho/pentaho-reporting/pull/1657/files#r1626727900

the other are just cleaning up the code

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java (hunk 1 above)
Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java (hunk 2 above)
Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java (hunk 3 above)
Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java (hunk 4 above)
Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java (hunk 5 above)
Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

File: engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java (hunk 6 above)
Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

Comment (soagarwal1) [reply engine/core/src/main/java/org/pentaho/reporting/engine/classic/core/modules/output/table/html/helper/HtmlTextExtractorHelper.java]: addressed

**Comment by buildguy** [comment]
<div align='center'>

[👍 Frogbot scanned this pull request and found that it did not add vulnerable dependencies.](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

<details>
<summary> <b>Note:</b> </summary>


---
<div align='center'>

**Frogbot** also supports **Contextual Analysis, Secret Detection, IaC and SAST Vulnerabilities Scanning**. This features are included as part of the [JFrog Advanced Security](https://jfrog.com/advanced-security) package, which isn't enabled on your system.

</div>


</details>


---
<div align='center'>

[🐸 JFrog Frogbot](https://docs.jfrog-applications.jfrog.io/jfrog-applications/frogbot)

</div>

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 8m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -Djs.no.sandbox -pl engine/core
```
---
:ok_hand: All tests passed!
 
Tests run: **822**, Failures: **0**, Skipped: **2** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/reporting/job/PR-1657/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by NJtwentyone — APPROVED** [review]

**Pull request merged**